Fix OrientationMap for Mesh

nilsvu opened the pull request:
## Proposed changes

The orientation map should take an object in the element and transform it to the neighbor's orientation. The logic of the OrientationMap::operator()(Mesh) was the reverse.

Reviewers: please check that I didn't mess this up.

This PR fixes a bug in the elliptic DG subdomain operator on the BNS domain (the BinaryCompactObject domain without excisions).

<s>To make sure that no existing code implicitly assumes the wrong behavior of the function, I made the old function emit an error, so any occurrences of it should be caught by the unit tests. I'll replace the old with the new function once we've confirmed all occurrences are updated.</s> done.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

kidder commented on src/Domain/Structure/OrientationMap.hpp:
```diff
@@ -70,6 +70,22 @@ class OrientationMap {
   /// The corresponding Mesh in the neighbor
   Mesh<VolumeDim> operator()(const Mesh<VolumeDim>& mesh) const;
 
+  /// The corresponding Mesh in the neighbor
```
I assume you commented out the above operator to make sure you changed all of the calls to it to your new function

nilsvu replied:
yep. I deleted the new function now, since I have verified that I caught all occurrences.

kidder commented on src/Domain/Structure/OrientationMap.hpp:
```diff
@@ -70,6 +70,22 @@ class OrientationMap {
   /// The corresponding Mesh in the neighbor
   Mesh<VolumeDim> operator()(const Mesh<VolumeDim>& mesh) const;
 
+  /// The corresponding Mesh in the neighbor
```
please add `Mesh` to the description of the class above

kidder reviewed (changes requested):
yep, looks like it is a bug to me

nilsvu committed:
Fix OrientationMap for Mesh

The orientation map should take an object in the element and transform it
to the neighbor's orientation. The logic of the
OrientationMap::operator()(Mesh) was the reverse.

mar-celine commented on src/Domain/Structure/OrientationMap.hpp:
```diff
@@ -120,6 +131,21 @@ std::array<T, VolumeDim> OrientationMap<VolumeDim>::permute_from_neighbor(
 
+template <size_t VolumeDim>
+template <typename T>
+std::array<T, VolumeDim> OrientationMap<VolumeDim>::permute_to_neighbor(
```
I think, just for consistency's sake, to have `permute_from_neighbor` precede `permute_to_neighbor` (or vice versa) in both the declaration and in the definition

nilsvu replied:
updated

mar-celine commented on src/Domain/Structure/OrientationMap.hpp:
```diff
@@ -120,6 +131,21 @@ std::array<T, VolumeDim> OrientationMap<VolumeDim>::permute_from_neighbor(
+std::array<T, VolumeDim> OrientationMap<VolumeDim>::permute_to_neighbor(
+    const std::array<T, VolumeDim>& array_to_permute) const {
+  std::array<T, VolumeDim> array_in_neighbor = array_to_permute;
+  if (is_aligned_ or VolumeDim <= 1) {
```
do we have `VolumeDim < 1` objects? this just stood out to me and made me wonder

nilsvu replied:
Yea meshes can have zero dim, e.g. when they represent the face of a 1d domain.

mar-celine replied:
I didn't know we had this capability! awesome :slightly_smiling_face:

mar-celine reviewed (commented):
Looks good!

nilsvu commented:
@kidder please take another look
